sale_commission/models/settlement.py

```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, exceptions, fields, models, _
+from odoo.exceptions import UserError
+
+
+class Settlement(models.Model):
+    _name = "sale.commission.settlement"
+    _description = "Settlement"
+
+    def _default_currency(self):
+        return self.env.user.company_id.currency_id.id
+
+    name = fields.Char('Name')
+    total = fields.Float(compute="_compute_total", readonly=True, store=True)
+    date_from = fields.Date(string="From")
+    date_to = fields.Date(string="To")
+    agent = fields.Many2one(
+        comodel_name="res.partner", domain="[('agent', '=', True)]")
+    agent_type = fields.Selection(related='agent.agent_type')
+    lines = fields.One2many(
+        comodel_name="sale.commission.settlement.line",
+        inverse_name="settlement", string="Settlement lines", readonly=True)
+    state = fields.Selection(
+        selection=[("settled", "Settled"),
+                   ("invoiced", "Invoiced"),
+                   ("cancel", "Canceled"),
+                   ("except_invoice", "Invoice exception")], string="State",
+        readonly=True, default="settled")
+    invoice = fields.Many2one(
+        comodel_name="account.invoice", string="Generated invoice",
+        readonly=True)
+    currency_id = fields.Many2one(
+        comodel_name='res.currency', readonly=True,
+        default=_default_currency)
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        required=True
+    )
+
+    @api.depends('lines', 'lines.settled_amount')
+    def _compute_total(self):
+        for record in self:
+            record.total = sum(x.settled_amount for x in record.lines)
+
+    @api.multi
+    def action_cancel(self):
+        if any(x.state != 'settled' for x in self):
+            raise exceptions.Warning(
+                _('Cannot cancel an invoiced settlement.'))
+        self.write({'state': 'cancel'})
+
+    @api.multi
+    def unlink(self):
+        """Allow to delete only cancelled settlements"""
+        if any(x.state == 'invoiced' for x in self):
+            raise exceptions.Warning(
+                _("You can't delete invoiced settlements."))
+        return super(Settlement, self).unlink()
+
+    @api.multi
+    def action_invoice(self):
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _('Make invoice'),
+            'res_model': 'sale.commission.make.invoice',
+            'view_type': 'form',
+            'target': 'new',
+            'view_mode': 'form',
+            'context': {'settlement_ids': self.ids}
+        }
+
+    def _prepare_invoice_header(self, settlement, journal, date=False):
+        invoice = self.env['account.invoice'].new({
+            'partner_id': settlement.agent.id,
+            'type': ('in_invoice' if journal.type == 'purchase' else
+                     'in_refund'),
+            'date_invoice': date,
+            'journal_id': journal.id,
+            'company_id': self.company_id.id,
```

Comment:
When using the wizard to create the invoices for the commissions already settled, in cases where more than one agent was associated to an invoice line an error occurs. I fixed it by changing this line to:  'company_id': settlement.company_id.id,